File: app/src/main/java/org/qp/android/questopiabundle/LibResult.java
package org.qp.android.questopiabundle;

import android.os.Parcel;
import android.os.Parcelable;

public class LibResult<T extends Parcelable> implements Parcelable {

    public static final Creator<LibResult<Parcelable>> CREATOR = new Creator<>() {
        @Override
        public LibResult<Parcelable> createFromParcel(Parcel in) {
            return new LibResult<>(in);
        }

        @Override
        public LibResult<Parcelable>[] newArray(int size) {
            return new LibResult[size];
        }
    };

    public T value;

    public LibResult(T value, Class<T> classType) {
        this.value = value;
    }

    protected LibResult(Parcel in) {
        this.value = in.readParcelable(getClass().getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(value, flags);
    }

    @Override
    public int describeContents() {
        return 0;
    }
}

File: app/src/main/java/org/qp/android/questopiabundle/LibResult.kt
package org.qp.android.questopiabundle

import android.os.Parcel
import android.os.Parcelable

data class LibResult<T : Parcelable?>(
    var value: T?
) : Parcelable {

    constructor(source: Parcel) : this(null) {
        this.value = source.readParcelable(javaClass.classLoader)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(value, flags)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<LibResult<Parcelable>> {
        override fun createFromParcel(source: Parcel): LibResult<Parcelable> {
            return LibResult(source)
        }

        override fun newArray(size: Int): Array<LibResult<Parcelable>?> {
            return arrayOfNulls(size)
        }
    }
}
